src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -948,18 +948,14 @@ private boolean reuseCachedHandle(boolean hasNewTypeDefinitions, boolean discard
 		
 		// New type definitions and existing cached handle reference then deregister cached handle.
 		if(hasNewTypeDefinitions) {
-			if (null != cachedPreparedStatementHandle && hasPreparedStatementHandle() && prepStmtHandle == cachedPreparedStatementHandle.getHandle()) {
-				cachedPreparedStatementHandle.removeReference();
-	            cachedPreparedStatementHandle.setIsExplicitlyDiscarded();
-			}
+			resetPrepStmtHandle();
```

Comment:
We do need to keep the previous code that does `cachedPreparedStatementHandle.removeReference`, in fact maybe code should be refactored so that call does `resetPrepStmtHandle`. If we don't remove the reference it will leak. It is clear that the main problem with the previous code was the call to `cachedPreparedStatementHandle.setIsExplicitlyDiscarded`, that should not be called here.

Reply:
I added call to `removeReference` back. This needs to be here. Thanks for catching that!

Also I saw another random issue with executeBatch() calls, that failed with `Could not find prepared statement with handle x`. We were able to reproduce this issue by load testing Statement tests and saw same pattern in all exceptions. After long investigations, I came up with a fix in my recent push. 

I made a change to `retryBasedOnFailedReuseOfCachedHandle` by passing `isBatch` parameter, which does not throw exception if a batch is being processed. There are intermittent failures if we do not add this, since if a handle is fetched from cache, and `needsPrepare=true` as a result of type def change, it throws exception while executing batch.

Let me know your thoughts on this. We have not seen any failure with this fix after running tests in multi-threaded environments with load testing.